iceoryx_dds/source/iceoryx_dds/dds/fastdds_data_reader.cpp

```diff
@@ -0,0 +1,287 @@
+{
+    if (m_dataReader != nullptr)
+    {
+        m_subscriber->delete_datareader(m_dataReader);
+    }
+    if (m_subscriber != nullptr)
+    {
+        m_participant->delete_subscriber(m_subscriber);
+    }
+    if (m_topic != nullptr)
+    {
+        m_participant->delete_topic(m_topic);
+    }
+    if (m_participant != nullptr)
+    {
+        eprosima::fastdds::dds::DomainParticipantFactory::get_instance()->delete_participant(m_participant);
+    }
+    LogDebug() << m_logName << "Destroyed FastDDSDataReader";
+}
+
+void iox::dds::FastDDSDataReader::connect() noexcept
+{
+    if (!m_isConnected.load(std::memory_order_relaxed))
+    {
+        // Create the DomainParticipant
+        eprosima::fastdds::dds::DomainParticipantQos participantQos;
+        m_participant =
+            eprosima::fastdds::dds::DomainParticipantFactory::get_instance()->create_participant(0, participantQos);
+
+        // Register the type
+        m_type.register_type(m_participant);
+
+        // Create topic
+        auto topicString =
+            "/" + std::string(m_serviceId) + "/" + std::string(m_instanceId) + "/" + std::string(m_eventId);
+        m_topic = m_participant->create_topic(topicString, "Mempool::Chunk", eprosima::fastdds::dds::TOPIC_QOS_DEFAULT);
+
+        // Create subscriber
+        m_subscriber = m_participant->create_subscriber(eprosima::fastdds::dds::SUBSCRIBER_QOS_DEFAULT, nullptr);
+
+        // Create datareader
+        eprosima::fastdds::dds::DataReaderQos dataReaderQos = eprosima::fastdds::dds::DATAREADER_QOS_DEFAULT;
+        dataReaderQos.history().kind = eprosima::fastdds::dds::KEEP_ALL_HISTORY_QOS;
+        dataReaderQos.endpoint().history_memory_policy =
+            eprosima::fastrtps::rtps::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
+        m_dataReader = m_subscriber->create_datareader(m_topic, dataReaderQos, &m_listener);
+
+        // Save state
+        m_isConnected.store(true, std::memory_order_relaxed);
+        LogDebug() << m_logName << "Connected to topic: " << topicString;
+    }
+}
+
+iox::cxx::optional<uint32_t> iox::dds::FastDDSDataReader::peekNextSize()
+{
+    // Get information about the next unread sample
+    Mempool::Chunk data{0, nullptr};
+    eprosima::fastdds::dds::SampleInfo nextInfo;
+    if (eprosima::fastrtps::types::ReturnCode_t::RETCODE_OK ==
+        m_dataReader->read_next_sample(&data ,&nextInfo))
+    {
+        if (data.size != 0)
+        {
+            return iox::cxx::optional<uint32_t>(static_cast<uint64_t>(data.size));
+        }
+    }
+
+    // no valid samples available
+    return iox::cxx::nullopt_t();
+}
+
+bool iox::dds::FastDDSDataReader::hasNewSamples()
+{
+    eprosima::fastdds::dds::SampleInfo nextInfo;
+    return (m_dataReader->get_first_untaken_info(&nextInfo) == eprosima::fastrtps::types::ReturnCode_t::RETCODE_OK);
+}
+
+iox::cxx::expected<iox::dds::DataReaderError> iox::dds::FastDDSDataReader::takeNext(
+        uint8_t* const buffer,
```

Comment:
In iceoryx_utils we have a type called `iox::cxx::not_null` which can be used to restrict pointer arguments to not be null.
You can find it in the header `iceoryx_utils/cxx/helplets.hpp`. It behaves like a normal pointer and could be used like
```cpp
iox::cxx::expected<iox::dds::DataReaderError> iox::dds::FastDDSDataReader::takeNext(
        iox::cxx::not_null<uint8_t*> buffer,  const uint64_t& bufferSize)
```

When using this you do not require any error handling like in line 124 `if ( buffer == nullptr )`.
If possible `cxx::not_null` tries to ensure this condition at compile time. You can find more info here: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Ri-nullptr